_posts/2022-01-21-safe-releases.md

```diff
@@ -0,0 +1,95 @@
+Database migrations
+Uploading build assets to external sources
+Preparing new environment configurations
+
+This is a danger zone because your data changes need to be backwards-compatible with both versions of code which may be running against your environment. Before the next step, your old code is still being run. After the deployment, your new code will be active.
+
+## Danger Zone 2 - Code Deploy Phase
+
+Finally, your new code is pushed to servers. This can happen in one go or it can be controlled with architectures like blue-green deployments. This is also a danger zone because the new code will fail if your release commands did not prepare the environment properly.
+
+
+## Additional Considerations
+
+Besides these two hot-spots, there are a few other things you should consider before kicking off any production deployments:
+
+### Is your release affecting a distributed system?
+
+Releasing changes to a single application is one thing, but if your system interacts with others and your changes could affect those interactions, you'll want to take special care.
+
+Below I outline a general process which can also be applied to multi-system releases, while noting that there will generally be many more releases involved in order to gracefully transition each system safely while it is still interacting with other apps.
+
+### Does your release contain long-running data migrations?
+
+One common request in a software business is to release a feature and prefill a bunch of data. Since Rails migrations support full Ruby and Rails code, it's tempting to just write a loop in there and populate a bunch of records.
+
+This is dangerous! Without knowing exactly how long a migration can take, you run the risk of timing out your release commands and failing the build. Depending on your configuration and whether you moved this outside of a transaction, this could leave you with an inconsistent DB state.
+
+### Does your release add indices, locks, or perform other heavy table operations?
+
+It's also important to be aware of expensive database operations like adding locks or indices, changing column types, etc.
+
+There are great resources available [1] to learn whether a particular action is safe or not, and as you learn more about them you will also learn strategies to release these changes without any downtime or user-facing side-effects.
+
+## 1-2-3: A slow, safe release process
+
+For the next piece, we'll take a simple scenario and describe the releases necessary for a safe migration. The request is: move the `email` column from the `user` table to the `user_details` table. Keep in mind that there are existing emails in the original table.
+
+My rule of thumb is: The more explicit, smaller steps, the safer the release (to a point!). The more you break things down into parts, the less likely any release will result in a broken build or put your application into a messy state that requires manual intervention.
```

Comment:
If you want, you can even call this out -- more steps doesn't always mean more work. Your CI/CD pipeline will take care of the repetative steps that are a result of more steps.